cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+		} else if err != nil {
+			return nil, fmt.Errorf("cannot read dm name for disk %s: %v", majorMinor, err)
+		}
+
+		dmUUIDFile := filepath.Join(dirs.SysfsDir, "dev/block/", majorMinor, "dm/uuid")
+		dmUUID, err := ioutil.ReadFile(dmUUIDFile)
+		if err != nil && os.IsNotExist(err) {
+			// not a dm device
+			return nil, fmt.Errorf("disk %s is not a decrypted device", majorMinor)
+		} else if err != nil {
+			return nil, fmt.Errorf("cannot read dm uuid for disk %s: %v", majorMinor, err)
+		}
+
+		// trim the suffix of the dm name from the dm uuid to safely match the
+		// regex - the dm uuid contains the dm name, and the dm name is user
+		// controlled, so we want to remove that and just use the luks pattern
+		// to match the device uuid
+		// we are extra safe here since the dm name could be hypothetically user
+		// controlled via an external USB disk with LVM partition names, etc.
+		dmUUIDSafe := bytes.TrimSuffix(
+			bytes.TrimSpace(dmUUID),
+			append([]byte("-"), bytes.TrimSpace(dmName)...),
+		)
+		matches := luksUUIDPatternRe.FindSubmatch(dmUUIDSafe)
+		if len(matches) != 2 {
+			// the format of the uuid is different - new luks version maybe?
+			return nil, fmt.Errorf(
+				"cannot verify disk: partition %s does not have a valid luks uuid format",
+				majorMinor,
+			)
+		}
+
+		// the uuid is the first and only submatch, but it is not in the same
+		// format exactly as we want to use, namely it is missing all of the "-"
+		// characters in a typical uuid, i.e. it is of the form:
+		// ae6e79de00a9406f80ee64ba7c1966bb but we want it to be like:
+		// ae6e79de-00a9-406f-80ee-64ba7c1966bb so we need to add in 4 "-"
+		// characters
+		fullUUID := string(matches[1])
+		realUUID := fmt.Sprintf(
+			"%s-%s-%s-%s-%s",
+			fullUUID[0:8],
+			fullUUID[8:12],
+			fullUUID[12:16],
+			fullUUID[16:20],
+			fullUUID[20:],
+		)
+
+		// now finally, we need to use this uuid, which is the device uuid of
+		// the actual physical encrypted partition to get the path, which will
+		// be something like /dev/vda4, etc.
+		props, err := udevProperties(filepath.Join("/dev/disk/by-uuid", realUUID))
+		if err != nil {
+			return nil, fmt.Errorf("cannot verify partition with uuid %s: %v", realUUID, err)
+		}
+
+		path, ok := props["DEVNAME"]
+		if !ok {
+			return nil, fmt.Errorf("cannot verify partition with uuid %s: incomplete udev information", realUUID)
+		}
+
+		// save it in the mountpointPart for the rest of the function to operate
+		// normally on
+		mountpointPart.path = path
+	}
+
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	props, err := udevProperties(mountpointPart.path)
+	if err != nil && props == nil {
+		// only fail here if props is nil, if it's available we validate it
+		// below
+		return nil, fmt.Errorf("cannot find disk for partition %s: %v", mountpointPart.path, err)
```

Comment:
probably some of the error cases (most expected ones?) need some testing

Reply:
well the reason we do best effort is because afaict this is a bit racy in the initrd, so I'm not sure which of these checks are expected to commonly fail, etc.

Reply:
what does that mean:

* that sometime the cross check will fail? and we'll need a reboot
* or that the sysfs for the bits we are not interested in might be not in a stable state?

I fear either way this might be a misleading API and we should make it less like it seems is capturing all the information and more goal oriented.